### PR TITLE
Encode / Decode

### PR DESCRIPTION
**Pull request by trstephen**
Done done done! ✅

### CODE REVIEW

File: wm_encoder.m [hunk 1]
```diff
@@ -11,28 +11,25 @@
...
+% Need make sure the message is < 199 mod 6 == 33 characters since
+% we'll be mapping characters to 6 bit binary sequences.
+bch_n = 255; bch_k = 199; word_size = 6;
+max_wm_len = floor(bch_k / word_size);
+
+wm_len = length(wm);
+if wm_len > max_wm_len
+    error('Error.\nMax watermark length is %d but your message is %d.', ...
+        max_wm_len, wm_len);
+end
```
Comment (houtan1) [inline wm_encoder.m:27]: Great word length check.

File: wm_encoder.m [hunk 2]
```diff
@@ -41,40 +38,67 @@
...
 
 % Convert the user inserted watermark phrase into lower case, then use the
 % map to convert it to a set of minimal offsets.
 wm = lower(wm);
 
-offsets = zeros(size(wm));
+% Convert the watermark phrase to a binary sequence
+% TODO: Preallocate space to satisfy warning.
+%       This is a bit tricky because I have to block-assign values
+%       to wm_bin_seq, which I don't think is possible??
```
Comment (houtan1) [inline wm_encoder.m:58]: I think we can live with this MATLAB warning.

File: wm_encoder.m [hunk 3]
```diff
@@ -41,40 +38,67 @@
...
 for i = 1:length(wm)
-    if letter_offsets.isKey(wm(i))
-        offsets(i) = letter_offsets(wm(i));
+    if letter_map.isKey(wm(i))
+        % Append the binary value of the character to the sequence
+        wm_bin_seq = [wm_bin_seq letter_map(wm(i))];
     else
-        offsets(i) = 0;
+        % Append a ' ' if the character isn't in our map
+        wm_bin_seq = [wm_bin_seq letter_map(' ')];
```
Comment (houtan1) [inline wm_encoder.m:66]: Graceful handling of edge cases.

File: wm_encoder.m [hunk 4]
```diff
@@ -41,40 +38,67 @@
...
     end
 end
 
+% We have a 6x(num chars) matrix that needs to be converted into a 1x(6 x num chars) vector
+wm_bin_seq = str2num(reshape(wm_bin_seq, 1, [])')';
+
+% Determine BCH correction code
+% Have to pad up to valid (n,k) tuples for BCH codes
+% See: https://www.mathworks.com/help/comm/ref/bchenc.html
```
Comment (houtan1) [inline wm_encoder.m:75]: Great commenting and references throughout.

File: .gitignore [hunk 5]
```diff
@@ -0,0 +1,2 @@
+Watermarked.tiff
```
Comment (houtan1) [inline .gitignore:1]: I would have made the rookie mistake of not gitignoring that file 😋

File: wm_decoder.m [hunk 6]
```diff
@@ -37,23 +33,53 @@
...
+% originally mapped '1' -> 2 and '0' -> 0 we want the unmapping to be
+% -1,0 -> '0' and 1, 2 -> '1'
+for i = 1:length(wm_offsets)
+    if wm_offsets(i) > 0
+        wm_offsets(i) = 1;
+    else
+        wm_offsets(i) = 0;
+    end
+end
+
```
Comment (houtan1) [inline wm_decoder.m:46]: 👍

File: wm_decoder.m [hunk 7]
```diff
@@ -37,23 +33,53 @@
...
+wm_dec = wm_bch_dec(1:(max_wm_len * word_size));
+
+% Group the elements in our decoded vector into 6bit binary words.
+% Convert the words to their decimal value for use in the dec -> char map.
+dec_offsets = zeros(1, max_wm_len);
+for i = 1:max_wm_len
+    first_bit = (i-1)*word_size + 1;
+    last_bit = first_bit + (word_size-1);
+    word = wm_dec(first_bit:last_bit);
+    dec_offsets(i) = bin2dec(dec2bin(word.x)'); % vector of 0/1 -> binary char vector -> decimal
```
Comment (houtan1) [inline wm_decoder.m:67]: Pure Magic 🎉

**Review by houtan1 — APPROVED** [review]
Reviewed your updates to wm_encoder.m and wm_decoder.m, ran them both on my machine with MATLAB R2016b and they behaved as expected. Good error checking on the message length of 33 chars or less!

**Comment by trstephen** [comment]
Resolves #1, #2, #3, #6

**Pull request merged**